### PR TITLE
Merge the OpenSsl DuplicateKeyHandle feature from dev-sslstream

### PR DESCRIPTION
**Pull request by bartonjs**
In order to work around a package sequencing problem we're pulling this feature into master ahead of the rest of dev-sslstream.

### CODE REVIEW

File: src/System.Security.Cryptography.OpenSsl/System.Security.Cryptography.OpenSsl.sln [hunk 1]
```diff
@@ -27,6 +32,20 @@ Global
 		{78452F3E-BA91-47E7-BB0F-02E8A5C116C4}.OSX_Debug|Any CPU.Build.0 = OSX_Debug|Any CPU
 		{78452F3E-BA91-47E7-BB0F-02E8A5C116C4}.OSX_Release|Any CPU.ActiveCfg = OSX_Release|Any CPU
 		{78452F3E-BA91-47E7-BB0F-02E8A5C116C4}.OSX_Release|Any CPU.Build.0 = OSX_Release|Any CPU
+		{A05C2EF2-A986-448C-9C63-735CC17409AA}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{A05C2EF2-A986-448C-9C63-735CC17409AA}.Debug|Any CPU.Build.0 = Debug|Any CPU
```
Comment (stephentoub) [inline src/System.Security.Cryptography.OpenSsl/System.Security.Cryptography.OpenSsl.sln]: Do we want/need these non-OS-specific entries?

File: src/System.Security.Cryptography.OpenSsl/src/System.Security.Cryptography.OpenSsl.csproj [hunk 2]
```diff
@@ -59,6 +66,12 @@
     <Compile Include="$(CommonPath)\Interop\Unix\libcrypto\Interop.EcDsa.cs">
       <Link>Common\Interop\Unix\libcrypto\Interop.EcDsa.cs</Link>
     </Compile>
+    <Compile Include="$(CommonPath)\Interop\Unix\System.Security.Cryptography.Native\Interop.Initialization.cs">
+      <Link>Common\Interop\Unix\System.Security.Cryptography.Native\Interop.Initialization.cs"</Link>
+    </Compile>
+    <Compile Include="$(CommonPath)\Interop\Unix\System.Security.Cryptography.Native\Interop.EvpPkey.cs">
+      <Link>Common\Interop\Unix\System.Security.Cryptography.Native\Interop.EvpPkey.cs</Link>
+    </Compile>
```
Comment (stephentoub) [inline src/System.Security.Cryptography.OpenSsl/src/System.Security.Cryptography.OpenSsl.csproj]: We have both System.Security.Cryptography.Native\Interop.EvpPkey.cs and libcrypto\Interop.EvpPkey.cs?

**Comment by bartonjs** [comment]
cc @stephentoub 

This is just #3403 in reverse (scoped to only the one feature).

File: src/System.Security.Cryptography.OpenSsl/System.Security.Cryptography.OpenSsl.sln (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.OpenSsl/System.Security.Cryptography.OpenSsl.sln]: Every time I try to delete them it just adds them back.

File: src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs [hunk 3]
```diff
@@ -41,25 +40,66 @@ public ECDsaOpenSsl(IntPtr handle)
...
+
+        /// <summary>
+        /// Obtain a SafeHandle version of an EVP_PKEY* which wraps an EC_KEY* equivalent
+        /// to the current key for this instance.
+        /// </summary>
+        /// <returns>A SafeHandle for the EC_KEY key in OpenSSL</returns>
+        public SafeEvpPKeyHandle DuplicateKeyHandle()
+        {
+            SafeEcKeyHandle currentKey = _key.Value;
+            SafeEvpPKeyHandle pkeyHandle = Interop.libcrypto.EVP_PKEY_new();
```
Comment (stephentoub) [inline src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs]: Do we need to validate IsInvalid?

File: src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/RSAOpenSsl.cs [hunk 4]
```diff
@@ -88,6 +119,35 @@ public override int KeySize
...
 
+        /// <summary>
+        /// Obtain a SafeHandle version of an EVP_PKEY* which wraps an RSA* equivalent
+        /// to the current key for this instance.
+        /// </summary>
+        /// <returns>A SafeHandle for the RSA key in OpenSSL</returns>
+        public SafeEvpPKeyHandle DuplicateKeyHandle()
+        {
+            SafeRsaHandle currentKey = _key.Value;
+            SafeEvpPKeyHandle pkeyHandle = Interop.libcrypto.EVP_PKEY_new();
```
Comment (stephentoub) [inline src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/RSAOpenSsl.cs]: Ditto on IsInvalid?

File: src/System.Security.Cryptography.OpenSsl/src/System.Security.Cryptography.OpenSsl.csproj (hunk 2 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.OpenSsl/src/System.Security.Cryptography.OpenSsl.csproj]: Yeah, Native has a function we needed that wasn't exposed; and the libcrypto one calls directly.

As the shimming continues the libcrypto one will go away.

**Comment by stephentoub** [comment]
A few questions/comments, but otherwise LGTM.

**Pull request merged**